experimental/StandardFiniteFields/src/StandardFiniteFields.jl

```diff
@@ -0,0 +1,509 @@
+        return 1
+      elseif k == 2
+        a = non_rth_root(standard_finite_field(p, 2), r)
+        # Xr[2]^2 -a
+        return steinitz_number(-a)
+      else
+        # Xr[i]^2 - Xr[i-1]
+        return (p - 1) * ZZ(p)^(r^(k - 2))
+      end
+    elseif r == 2
+      if k == 1
+        # Xr[1]^2 -a
+        a = non_rth_root(standard_finite_field(p, 1), r)
+        return steinitz_number(-a)
+      else
+        # Xr[j]^r - Xr[j-1]
+        return (p - 1) * ZZ(p)^(r^(k - 2))
+      end
+    else
+      # Here we use pseudo-random polynomials...
+      F = standard_finite_field(p, r^(k - 1))
+      if k == 1
+        a = -one(F)
+      else
+        a = -gen(F)
+      end
+      l = standard_irreducible_coefficient_list(F, r, a)
+      pop!(l)
+      while is_zero(l[end])
+        pop!(l)
+      end
+      return evaluate(polynomial(ZZ, steinitz_number.(l)), order(F))
+    end
+  end
+end
+
+# x will be represented internally as a polynomial over Fp in the generator of F.
+# We need to first convert this to an Fp-vector, then to the corresponding vector
+# with respect to the Tower Basis.
+# Then we think of this vector as a polynomial (over ZZ) in a temporary indeterminate z,
+# and evaluate at z = char(F) to get the Steinitz number.
+# NOTE for whatever reason, evaluate(polynomial(), ) is faster than evalpoly()
+function steinitz_number(F::PrimeField, x::PrimeFieldElem)
+  @assert parent(x) === F
+  return lift(x)
+end
+function steinitz_number(F::FinField, x::FinFieldElem)
+  @assert parent(x) === F
+  v = lift.(absolute_coordinates(x) * primitive_powers_in_tower_basis(F))
+  return evaluate(polynomial(ZZ, v), characteristic(F))
+end
+function steinitz_number(x::FinFieldElem)
+  return steinitz_number(parent(x), x)
+end
+
+# describes monomials in tower basis plus their degrees
+function standard_monomial(n::IntegerUnion)
+  error("not implemented")
+end
+# just returns degrees of monomials in tower basis
+# TODO : pass in factorization? Do we need this with memoization?
+function standard_monomial_degrees(n::IntegerUnion)::Vector{Int}
+  if n == 1
+    return [1]
+  end
+  # need the largest prime factor a of n
+  a, k = largest_factor(n)
+  res = standard_monomial_degrees(div(n, a))
+  m = a^k
+  new = map(x -> lcm(x, m), res)
+  for i = 1:a-1
+    append!(res, new)
+  end
+  return res
+end
+# map of monomials for degree n -> monomials of degree m by positions
+function standard_monomial_map(n::IntegerUnion, m::IntegerUnion)
+  d = standard_monomial_degrees(m)
+  return [i for i = 1:length(d) if mod(n, d[i]) == 0]
+end
+
+# Embed an element x of Fp^n into Fp^m by Steinitz numbers
+# where nr = steinitz_number(Fp^n, x)
+# I hate hate hate these variable names copied (mostly) from Lübeck
+function embed_steinitz(p::IntegerUnion, n::IntegerUnion, m::IntegerUnion, nr::IntegerUnion)
+  if n == m || iszero(nr)
+    return nr
+  end
+  l = digits(nr, base = Int(p))
+  m = @view standard_monomial_map(n, m)[1:length(l)]
```

Comment:
this function is not typesafe, as `m` first is an `IntegerUnion`, and now some kind of container. Maybe use two distinct names to distinguish?